bitmind/synthetic_image_generation/synthetic_image_generator.py

```diff
@@ -1,29 +1,201 @@
-            
-            img_filename = f"{save_dir}/{safe_prompt}-{index}.png"
-            return img, img_filename
\ No newline at end of file
+
+    def __init__(
+        self,
+        prompt_type='random',
+        prompt_generator_name=PROMPT_GENERATOR_NAMES[0],
+        diffuser_name=DIFFUSER_NAMES[0],
+        use_random_diffuser=False,
+        image_cache_dir=None
+    ):
+        assert prompt_type in PROMPT_TYPES,\
+            f'Invalid prompt type. Options are {prompt_type}'
+        assert prompt_generator_name in PROMPT_GENERATOR_NAMES,\
+            f'Invalid prompt generator name. Options are {PROMPT_GENERATOR_NAMES}'
+        assert use_random_diffuser or diffuser_name in DIFFUSER_NAMES, \
+            f'Invalid diffuser name. Options are {DIFFUSER_NAMES}'
+
+        self.use_random_diffuser = use_random_diffuser
+        self.prompt_type = prompt_type
+        self.prompt_generator_name = prompt_generator_name
+
+        if self.use_random_diffuser and diffuser_name is not None:
+            bt.logging.warning("Warning: diffuser_name will be ignored (use_random_diffuser=True)")
+            self.diffuser_name = None
+        else:
+            self.diffuser_name = diffuser_name
+
+        self.image_annotation_generator = None
+        if self.prompt_type == 'annotation':
+            bt.logging.info(f"Loading image captioning model ({IMAGE_ANNOTATION_MODEL})...")
+            self.image_annotation_generator = ImageAnnotationGenerator(model_name=IMAGE_ANNOTATION_MODEL)
+        else:
+            bt.logging.info(f"Loading prompt generation model ({prompt_generator_name})...")
+            self.prompt_generator = pipeline(
+                'text-generation', **PROMPT_GENERATOR_ARGS[prompt_generator_name])
+
+        self.image_cache_dir = image_cache_dir
+        if image_cache_dir is not None:
+            os.makedirs(self.image_cache_dir, exist_ok=True)
+
+        if diffuser_name is not None:
+            bt.logging.info(f"Loading image generation model ({diffuser_name})...")
+            self.diffuser = DiffusionPipeline.from_pretrained(
+                diffuser_name, torch_dtype=torch.float16, **DIFFUSER_ARGS[diffuser_name])
+            self.diffuser.to("cuda")
+        else:
+            bt.logging.info("A random image generation model will be loaded on each generation step.")
+            self.diffuser = None
+
+    def generate(self, k: int = 1, real_images=None) -> list:
+        """
+        Generates k synthetic images. If self.prompt_type is 'annotation', a BLIP2 captioning pipeline is used
+        to produce a prompt by captioning a real image. If self.prompt_type is 'random', an LLM is used to generate
+        a random prompt.
+
+        Args:
+            k (int): Number of images to generate.
+
+        Returns:
+            list: List of dictionaries containing 'prompt', 'image', and 'id'.
+        """
+        bt.logging.info("Generating prompts...")
+        if self.prompt_type == 'annotation':
+
```

Comment:
suggestion: We should assert that real_images is not None if prompt_type is "annotation"

Reply:
Addressed - used an exception instead because it felt right, then read this short article and it felt even more right https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/

Reply:
replaced the asserts in __init__ with exceptions as well